In [27]:
# Write a File 

f=open("demo.txt","w")
f.write("Hello")
f.close()

In [20]:
print(type(f))

<class '_io.TextIOWrapper'>


In [10]:
f=open("demo.txt","r")
data=f.read()
print(data)
f.close()


Hello Python


In [7]:
import os 
path=os.getcwd()

In [12]:
import os
print(os.getcwd())

/Users/anujshahdeo/Documents/GithubRepos/DE_Python_AWS/File Handling


In [13]:
print(os.listdir())

['out.txt', 'demo.txt', 'sample_1.csv', 'File_Handling.ipynb', 'sample.json', 'Notes.md', 'sample_out.csv', 'out.json', 'sample.txt']


In [15]:
os.path.exists("demo1.txt")

False

In [16]:
os.path.isfile("demo.txt")

True

In [17]:
os.path.isdir("demo")

False

In [19]:
os.path.join("File Handing","data.txt")

'File Handing/data.txt'

In [26]:
if  os.path.exists("demo.txt"):
    f=open("demo.txt","r")



In [35]:


# Output: 10
def second_highest(nums):
    first=second=float('-inf')
    for num in nums:
        if num > first:
            second =first
            first=num
        elif num > second and num != first:
            second=num
    
    return second if second != float('-inf') else None

print(second_highest([10,20,5,8,20]))


10


In [30]:
nums = [10,20,5,8,20]
second_highest(nums)

In [31]:
print(-999999 > float('-inf'))

True


In [42]:
try:
    with open("Notes.md","r") as f1,open("sample_1.csv","w") as f2:
        data=f1.read()
        #print(data)
        #f2.write(data)
except FileNotFoundError:
    print("FIle not Found")


In [ ]:
from datetime import datetime
INPUT_FILE="students.txt"
OUTPUT_FILE="output.txt"
LOG_FILE="app.log"

def write_log(level,message):
    timestamp=datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    log_message=f"{timestamp}, {level}, {message}\n"
    with open(LOG_FILE,"a",encoding="utf-8") as log_file:
        log_file.write(log_message)

def get_marks(marks):
    if marks>=90:
        return"A"
    elif marks >=75:
        return "B"
    elif marks >=40:
        return "C"
    else:
        return "Fail"
def get_result(marks):
    return "Pass" if marks>=40 else "Fail"

def process_students():
    total_records=0
    success_records=0
    failed_records=0

    write_log("INFO","Process Starated")

    try:
        with open(INPUT_FILE,"r",encoding="utf-8") as infile,open(OUTPUT_FILE,"w",encoding="utf-8") as outfile:
            for line_number,line in enumerate(infile,start=1):
                total_records+=1
                line=line.strip()

                if not line:
                    write_log("WARNING",f"Skipping empty line at line {line_number}")
                    continue
                write_log("INFO",f"Reading Line {line_number}")
                parts=line.split(",")

                if len(parts) !=3:
                    failed_records+=1
                    write_log("ERROR",f"Invalid format at line {line_number}:{line}")
                    continue
                student_id,name,marks=parts

                try:
                    marks=int(marks)
                except ValueError:
                    failed_records+=1
                    write_log("ERROR",f"Invalid marks at line {line_number}:{line}")
                    continue
                result=get_result(marks)
                grade=get_marks(marks)
                output_line=f"{student_id},{name},{marks},{result},{grade}\n"
                outfile.write(output_line)
                success_records+=1
                write_log("INFO",f"Successfully processed student_id={student_id}")
        write_log("INFO",f"Process Completed. Total={total_records},success={success_records},Failed={failed_records}") 
        print("Processing completed successfully")  
        print(f"Total records:{total_records}")
    except FileNotFoundError:
        write_log("ERROR",f"Input File not found: {INPUT_FILE}")    
        print(f"Error: Input file {INPUT_FILE} not found")
    except Exception as e:
        write_log("ERROR",f"Unexpected Error: {str(e)}")    
        print(f"Error",f"Unexpected Error : {str(e)}")

if __name__=="__main__":
    process_students()

In [45]:
process_students()

Processing completed successfully
Total records:7


In [55]:
with open(INPUT_FILE,"r",encoding="utf-8") as infile,open(OUTPUT_FILE,"w",encoding="utf-8") as outfile:
            for line_number,line in enumerate(infile,start=1):
                    print(line_number)
                    
                    line=line.strip()
                    #print(line)
                    parts=line.split(",")
                    print(parts)
                    student_id,name,marks=parts
                #     if len(parts) !=3:
                #         print(parts)

1
['student_id', 'name', 'marks']
2
['101', 'Rahul', '78']
3
['']


ValueError: not enough values to unpack (expected 3, got 1)

In [76]:
# Module 6 

with open("employee.txt","r",encoding="utf-8") as f:
    data=f.read(6) # reads file content 
    print(data)

101,Jo


In [80]:
with open("employee.txt","r",encoding="utf-8") as f:
    line1=f.readline().strip()
    line2=f.readline().strip()
    line3=f.readline().strip()
    print(line1)
    print(line2)
    print(line3)

101,John,5000
102,Alice,7000
103,Bob,abc


In [81]:
with open("employee.txt","r",encoding="utf-8") as f:
    lines=f.readlines()
    print(lines)

['101,John,5000\n', '102,Alice,7000\n', '103,Bob,abc\n', '104,Emma,8000\n', '105,David,3000']


In [88]:
# 7 Write 

with open ("file_name.txt","w",encoding="utf-8") as f:
        f.write("hello python\n")
        f.write("Welcome to file handling in python")
        f.write("This is a new line\n")
        f.write("This is another line")

In [89]:
names=["alice","bob","charlie"]
with open ("file_name.txt","w",encoding="utf-8") as f:
    for name in names:
        f.write(name+"\n")

In [90]:
employee=[
    {"id":101,"name":"john","salary":5000},
    {"id":102,"name":"Alice","salary":9000}
]
with open ("file_name.txt","w",encoding="utf-8") as f:
    for emp in employee:
        line=f"{emp['id']},{emp['name']},{emp['salary']}\n"
        f.write(line)

In [95]:
names=["alice","bob","charlie"]
lines=[name+"\n" for name in names]
print(lines)
with open ("file_name.txt","w",encoding="utf-8") as f:
    f.writelines(lines)


['alice\n', 'bob\n', 'charlie\n']


In [96]:
names=["alice","bob","charlie"]
result="\n".join(names)
print(result)

alice
bob
charlie


In [ ]:
with open ("file_name.txt","w",encoding="utf-8") as f:
    f.writelines("\n".join(names)+ "\n") # slow 


In [99]:
with open ("file_name.txt","a",encoding="utf-8") as f:
    f.write("new data\n")

In [106]:
with open ("file_name.txt","r",encoding="utf-8") as f:
    print(f.read(8))
    print(f.tell())

alice
bo
8


In [108]:
with open ("file_name.txt","r",encoding="utf-8") as f:
    f.seek(8)
    print(f.read(4))

b
ch


In [ ]:
with open ("file_name.txt","w",encoding="utf-8") as f:
   f.write("abc")
   print(f.tell())
   f.write("def")
   print(f.tell())

7
10


In [140]:
with open ("file_name.txt","rb") as f:
    #print(f.read(5))
    f.seek(-5,2)
    print(f.read())
   


b'World'


In [3]:
with open ("file_name.txt","r+") as f:
    #f.write("New Hello")
    data=f.read(5)
    print(data)
    f.write("123")


New H


In [4]:
with open ("file_name.txt","w+") as f:
    f.write("Hello World")
    f.seek(0)
    data=f.read()
    print(data)

Hello World


In [6]:
with open ("file_name.txt","a+") as f:
    f.write("Appending new data\n")
    f.seek(0)
    f.write("New Hello\n")

employees.txt
101,John,5000
102,Alice,7000
103,Bob,abc
104,Emma,8000
105,David,3000


[
    {i"d":101,"name":John,"salary":5000},
    {}
]

with open :
    for line_no,line in enumurate(f,start=1):
        line=line.strip()
        parts line

Read file line by line
Store valid employees in a list of dictionaries
Validate salary (must be integer)
Apply bonus:
salary >= 7000 → 20%
salary >= 5000 → 10%
else → 5%
Write output to salary_output.txt
Log errors and processing steps in app.log


In [68]:
employee=[]
with open("employee.txt","r",encoding="utf-8") as f:
    for line_no,line in enumerate(f,start=1):
        line=line.strip()
        #print(line)
        parts=line.split(",")
        #print(parts)
        emp_id,name,salary=parts
        emp_dict={
            "id":int(emp_id),
            "name":name,
            "salary":salary
                    }
        
        employee.append(emp_dict)
print(employee)
        
        

[{'id': 101, 'name': 'John', 'salary': '5000'}, {'id': 102, 'name': 'Alice', 'salary': '7000'}, {'id': 103, 'name': 'Bob', 'salary': 'abc'}, {'id': 104, 'name': 'Emma', 'salary': '8000'}, {'id': 105, 'name': 'David', 'salary': '3000'}]


system.log

INFO User login
ERROR Database failed
INFO File uploaded
ERROR Timeout occurred
WARNING Disk space low
ERROR API failed

Read log file
Count:
INFO
ERROR
WARNING
Store counts in dictionary
Write summary to summary.txt
Write detailed logs to app.log

{
    "INFO":2,
    "ERROR":3,
    "WARNING:1
}

In [71]:
log_count={
    "INFO":0,
    "ERROR":0,
    "WARNING":0
}

with open("system.log","r",encoding="utf-8") as f:
    for line in f:
        line=line.strip()
        #print(line)
        for key in log_count:
            if key in line:
                log_count[key] +=1
print(log_count)


{'INFO': 2, 'ERROR': 3, 'WARNING': 1}


In [56]:
line="101,John,5000"
line=line.split(",")

In [57]:
line

['101', 'John', '5000']

In [58]:
emp_id,name,salary=line

In [59]:
emp_id

'101'

In [60]:
name

'John'

In [61]:
emp_dict={
    "id":int(emp_id),
    "name":name,
    "salary":salary
}

In [62]:
emp_dict

{'id': 101, 'name': 'John', 'salary': '5000'}

In [63]:
employee=[]
employee.append(emp_dict)

In [64]:
employee

[{'id': 101, 'name': 'John', 'salary': '5000'}]